### PR TITLE
Add generic HTTP and HTTPS streaming support.

### PR DESCRIPTION
**Pull request by robottwo**
Adds support for opening vanilla HTTP and HTTPS addresses.
Supports efficient streaming, gzip and bz2 compression,
as well as Kerberos and username/password (basic) http
authentication.

### CODE REVIEW

File: smart_open/smart_open_lib.py [hunk 1]
```diff
@@ -573,23 +588,6 @@ def __enter__(self):
     def __exit__(self, type, value, traceback):
         pass
 
-
-def make_closing(base, **attrs):
```
Comment (piskvorky) [inline smart_open/smart_open_lib.py]: Did we drop support for Python 2.6? Why this change?

**Comment by piskvorky** [comment]
Oh wow, that's really cool! Interesting feature too.

What is your use-case for this, how do you use HTTP streaming yourself, if you don't mind me asking?

File: smart_open/smart_open_lib.py (hunk 1 above)
Comment (robottwo) [inline smart_open/smart_open_lib.py]: Oops, sorry, didn't think those comments were still active since python 2.7 has been around so long and there's no reason I know of for people to stick with 2.6. I can revert this part if you need 2.6 support -- let me know.

**Comment by robottwo** [comment]
I don't have a specific use-case for http streaming specifically (although I can imagine some) -- it's more that I like the idea of being able to use smart_open as a unified go-to tool for any type of uri. I definitely do have use-cases where I want to pull a web page, or a compressed file off of an http server.

File: smart_open/smart_open_lib.py (hunk 1 above)
Comment (tmylk) [inline smart_open/smart_open_lib.py]: We still support 2.6

File: smart_open/smart_open_lib.py [hunk 2]
```diff
@@ -604,15 +602,136 @@ def file_smart_open(fname, mode='rb'):
...
+
+    """
+    def __init__(self, url, mode='r', kerberos=False, user=None, password=None):
+        """
+        If Kerberos is True, will attempt to use the local Kerberos credentials.
+        Otherwise, will try to use "basic" HTTP authentication via username/password.
+
+        If none of those are set, will connect unauthenticated.
+        """
+        PY2 = sys.version_info[0] == 2
```
Comment (tmylk) [inline smart_open/smart_open_lib.py]: The python 2 check is now used in two places. Better refactor to a global.

File: smart_open/smart_open_lib.py [hunk 3]
```diff
@@ -604,15 +602,136 @@ def file_smart_open(fname, mode='rb'):
...
+    if parsed_uri.scheme not in ('http', 'https'):
+        raise TypeError("can only process http/https urls")
+    if mode not in ('r', 'rb'):
+        raise NotImplementedError('Streaming write to http not supported')
+
+    url = parsed_uri.uri_path
+
+    response = HttpReadStream(url, **kwargs)
+
+    if url.endswith('.gz'):
```
Comment (tmylk) [inline smart_open/smart_open_lib.py]: the gz and bz2 handling code is identical to `file_smart_open`. Please refactor to avoid duplication.

File: smart_open/tests/test_smart_open.py [hunk 4]
```diff
@@ -168,6 +168,28 @@ def test_webhdfs_read(self):
...
+            content = (smart_open_object.read(3).decode('utf-8') + 
+                       smart_open_object.read(len(expected_str) - 3).decode('utf-8'))
+            self.assertEqual(content[:len(expected_str)], expected_str)
+
+    def test_https_readline(self):
+        """Can we stream from an HTTPS site correctly"""
+        with smart_open.HttpOpenRead(smart_open.ParseUri('https://www.google.com'), 'r') as smart_open_object:
+            content = smart_open_object.readline().decode('utf-8')
+            expected_str = '<!doctype html>'
+            self.assertEqual(content[:len(expected_str)], expected_str)
```
Comment (tmylk) [inline smart_open/tests/test_smart_open.py]: Please add tests for authentication using `requests_mock`. It would also be great if the tests worked without access to internet and `google.com`.

**Comment by tmylk** [comment]
Thanks for the new feature! Just minor refactoring and better tests with mocking are needed before merging.

File: smart_open/smart_open_lib.py (hunk 1 above)
Comment (robottwo) [inline smart_open/smart_open_lib.py]: OK, reverted in latest.

File: smart_open/smart_open_lib.py (hunk 2 above)
Comment (robottwo) [inline smart_open/smart_open_lib.py]: Refactored in latest.

File: smart_open/smart_open_lib.py (hunk 3 above)
Comment (robottwo) [inline smart_open/smart_open_lib.py]: Good Point. This will also make it easier to add support for other compression types later.

Done in latest version.

File: smart_open/tests/test_smart_open.py (hunk 4 above)
Comment (robottwo) [inline smart_open/tests/test_smart_open.py]: OK, fixed. I added a test for Basic http auth, and made the other tests not actually go to google.com. I didn't add a Kerbeos test because I'm not sure that's easily doable without running within a Kerborized environment.

**Comment by robottwo** [comment]
Thanks. I've addressed all the comments in the latest version of the PR. Please take a look.

Thanks,
-Dan

File: smart_open/tests/test_smart_open.py [hunk 5]
```diff
@@ -168,6 +168,31 @@ def test_webhdfs_read(self):
         smart_open_object = smart_open.WebHdfsOpenRead(smart_open.ParseUri("webhdfs://127.0.0.1:8440/path/file"))
         self.assertEqual(smart_open_object.read().decode("utf-8"), "line1\nline2")
 
+    @responses.activate
+    def test_http_read(self):
+        """Does webhdfs read method work correctly"""
```
Comment (tmylk) [inline smart_open/tests/test_smart_open.py]: Just a type in the docstring ""Does http read method work correctly"""

File: smart_open/tests/test_smart_open.py [hunk 6]
```diff
@@ -168,6 +168,31 @@ def test_webhdfs_read(self):
...
+        responses.add(responses.GET, "http://127.0.0.1/index.html", body='line1\nline2')
+        smart_open_object = smart_open.HttpOpenRead(smart_open.ParseUri("http://127.0.0.1/index.html"))
+        self.assertEqual(smart_open_object.read().decode("utf-8"), "line1\nline2")
+
+    @responses.activate
+    def test_https_readline(self):
+        """Does webhdfs read method work correctly"""
```
Comment (tmylk) [inline smart_open/tests/test_smart_open.py]: Typo in docstring as well

**Comment by tmylk** [comment]
Thanks a lot for the code changes!
May I ask for some cosmetic changes - an example of the new functionality in the README and a note in the changelog? 
Then will merge and release this week.

File: smart_open/tests/test_smart_open.py [hunk 7]
```diff
@@ -168,6 +168,31 @@ def test_webhdfs_read(self):
...
+        responses.add(responses.GET, "https://127.0.0.1/index.html", body='line1\nline2')
+        smart_open_object = smart_open.HttpOpenRead(smart_open.ParseUri("https://127.0.0.1/index.html"))
+        self.assertEqual(smart_open_object.readline().decode("utf-8"), "line1")
+
+    @responses.activate
+    def test_http_pass(self):
+        """Does http authentication work correctly"""
+        responses.add(responses.GET, "http://127.0.0.1/index.html", body='line1\nline2')
+        smart_open_object = smart_open.HttpOpenRead(smart_open.ParseUri("http://127.0.0.1/index.html"),
+                                                    user='me', password='pass')
```
Comment (piskvorky) [inline smart_open/tests/test_smart_open.py]: nitpick: we use hanging (not vertical) indent.

Comment (robottwo) [inline smart_open/tests/test_smart_open.py]: Ok, fixed

**Comment by robottwo** [comment]
All comments addressed in the latest version.
